### PR TITLE
Move scss files beside components

### PR DESCRIPTION
**Pull request by phess101**
Moved scss file related to atomic component inside of the component's folder. Allows for direct importing of relevant style, vs importing the entire css file into the project, that bloats the project's size. Also contains the styles for a single component with the markup, further clarifying the location and usage of the component styles, making maintenance easier.

- [ ] Related GitHub issue(s) linked in PR description
- [x] Destination branch merged, built and tested with your changes
- [x] Code formatted and follows best practices and patterns
- [x] Code builds cleanly (no _additional_ warnings or errors)
- [x] Manually tested
- [x] Automated tests are passing
- [x] No _decreases_ in automated test coverage
- [x] Documentation updated (readme, docs, comments, etc.)
- [x] Localization: No hard-coded error messages in code files (_minimally_ in string constants)
- [x] New component and/ or properties? Make sure to update storybook

### CODE REVIEW

File: src/atoms/buttons/button.scss
Comment (mrjones2014) [inline src/atoms/buttons/button.scss:1]: thought: consider creating a `_global` partial SCSS file with these imports, then you'd only need one `@import "../../_global";` line.

File: src/atoms/buttons/button.tsx
Comment (mrjones2014) [inline src/atoms/buttons/button.tsx:6]: 👏 I love it!

**Review by mrjones2014 — CHANGES_REQUESTED** [review]

File: src/atoms/buttons/button.scss
Comment (phess101) [reply src/atoms/buttons/button.scss]: Love the idea @mrjones2014. I added it in a later commit

File: src/assets/scss/_global.scss
Comment (wintondeshong) [inline src/assets/scss/_global.scss:1]: Thought: `andculturecode-javascript-react-components.scss` is more of a true "global" import file. Definitely up for suggestions, but recommend renaming and moving this file to ...

`assets/scss/6-components/_component-base.scss` with potential contents like so...

```scss
/*------------------------------------*\
  Component base SCSS import file
  -------------------------------
  Imported directly into individual
  component SCSS files. Anything added
  is imported into every component.
\*------------------------------------*/
@import "0-vendor/__vendor";
@import "1-settings/__settings";
@import "2-tools/__tools";
@import "3-generic/__generic";
```

**Review by wintondeshong — CHANGES_REQUESTED** [review]
@phess101 thanks for putting this proposal together! It does a good job of demonstrating how this works when applied to the rest of the repository's components.

There are a few discussions points below that'd I'd love to get your input on as well as other senior engineers. As for next steps, I'm going to share this PR with some other seniors to make sure we are all good with proceeding in this direction.

Once we get approval across the board, the next step will be for you to complete the PR by rolling it through the rest of the components. When that comes, please make sure to verify there aren't any warnings (right now there are quite a few) and that none of the components have visually or functionally regressed. Recommend doing a side-by-side with our current released version https://andculturecode.github.io/AndcultureCode.JavaScript.React.Components/?path=/story/*

**Review by joshuapeters — APPROVED** [review]
Assuming that all of the other points @wintondeshong brought up are addressed, this seems fine.

My thoughts on this are the following:

- I'm fairly impartial to the importing styles vs. global stylesheet discussion, but because this is a different pattern than we've used I think that it should be something that we all sign off on. Maybe throw a poll in the senior engineering channel?

- Is there a way to set up module aliasing in SASS/Typescript to reduce the amount of pathing needed to import files? Think edio as a reference how we could import things by doing "@core/import/the-thing". It would be great if we could do that for the other stylesheets so that if a file is moved it doesn't regress a ton of components (a big bonus for the global stylesheet)

update: it looks like you can define module aliases in webpack for all things in general.  [See here for reference.](https://webpack.js.org/configuration/resolve/)

**Comment by KevinBusch** [comment]
If we're going to move in this direction (which I'm totally on-board with BTW), I think we need to make it clear how we should refactor and how we should be using the folder structure within `/src/assets/scss`.  My initial thoughts are that we should be moving any functions, mixins, or anything else that can be reused by our component SCSS files into their own folder structure for reuse by the components.  The remaining SCSS file in `/src/assets/scss` are really things that are needed by the site which, in this case, is used by the Storybook UI.  If we can kill any unneeded SCSS files and document in that folder structure that shared SCSS should not belong there, I think that'll go a long way in helping people understand that the ITCSS structure within this repo is for this site and not for the individual components.

**Review by brandongregoryscott — APPROVED** [review]
While I'm fine with this direction if that's how the rest of the team would like to go, I'd probably prefer to keep it as-is, as it's what we typically do in projects, and I don't see a huge benefit in changing it. It feels a bit awkward to have to import the 'global' or 'base' classes in each file now - something we got for free before

File: src/assets/scss/_global.scss
Comment (KevinBusch) [reply src/assets/scss/_global.scss:1]: I can get on board with this, but do we really want to import `3-generic/__generic` ?  Is this really something we want to do?  Consider a project that uses a single react component from this repo.  By now using it, we've now overriden all the consumer's default html and body element styling.  I feel that's something that is something that should still be set on a per project basis.

**Comment by KevinBusch** [comment]
> While I'm fine with this direction if that's how the rest of the team would like to go, I'd probably prefer to keep it as-is, as it's what we typically do in projects, and I don't see a huge benefit in changing it. It feels a bit awkward to have to import the 'global' or 'base' classes in each file now - something we got for free before

Read my response to @wintondeshong 's comment above to get my initial thoughts.  That might change your mind slightly.  I think the biggest benefit with a component based SCSS approach is that allow the compiled CSS to be smaller for the consumer.  If this library of components grows exponentially, why should a consumer have all the extra CSS styling if the consumer is only using a single component from this repo?

**Comment by dylanjustice** [comment]
No protest here. Only way I could see myself getting tripped up is when I have multiple stylesheets for a single react component.

**Review by HeyKos — APPROVED** [review]
This direction works for me if it's the way the group wants to take it.

**Comment by myty** [comment]
I'll go ahead and leave my two cents here.  My optimistic side really likes this approach and the idea of keeping similar functionality close to each other (it lends well to vertical slice architecture concepts), however, my pessimistic side wonders it would be smart to at least find a way of "trying" this approach before committing to doing it across the board.  Are we giving ourselves an unneeded maintenance burden if it turns out we don't like it?  If it's not that big of deal to switch back, then I say, let's try it.

**Review by myty — APPROVED** [review]

File: src/assets/scss/_global.scss
Comment (wintondeshong) [reply src/assets/scss/_global.scss:1]: Definitely hear what you are saying. Overriding defaults is one of the "challenges" regardless of our approach to placement of SCSS files. The current implementation directs consumers to load the top-level/global SCSS file (andculturecode-javascript-react-components) to get _everything_ and from there consumers override parts.

Regardless of which direction we take, the future for this repo's provided vendor, settings, tools, generic, etc... will likely need adjusted and further reduced for the masses. There are aspects of __generic that we setup almost every time that _would_ benefit us. While yes, in a literal sense it only serve storybook for _this_ repo, these styles are not just intended for storybook. Our boilerplate and other consuming projects use this for both the application itself and storybook.

**Comment by wintondeshong** [comment]
> If we're going to move in this direction (which I'm totally on-board with BTW), I think we need to make it clear how we should refactor and how we should be using the folder structure within `/src/assets/scss`. My initial thoughts are that we should be moving any functions, mixins, or anything else that can be reused by our component SCSS files into their own folder structure for reuse by the components. The remaining SCSS file in `/src/assets/scss` are really things that are needed by the site which, in this case, is used by the Storybook UI. If we can kill any unneeded SCSS files and document in that folder structure that shared SCSS should not belong there, I think that'll go a long way in helping people understand that the ITCSS structure within this repo is for this site and not for the individual components.
> 
> Baby steps of course with this PR, but we shouldn't lose site of the ITCSS folder structure as we refactor all the things.

I'm totally down for clarifying whatever might not be clear. That said, the only refactor intended is to add import statements to the components and literally take the SCSS files that are _currently_ under the ITCSS structure's `6-components` and put them parallel to their respective component file. Idea being that none of the rest of the structure changes and thus really needs re-explaining (or so was the hope).

As I responded above, the further down layers of the triangle are intended for use by all applications. From there storybook implementations (even this library's storybook via `.storybook/storybook.css` can override).

**Comment by wintondeshong** [comment]
> Assuming that all of the other points @wintondeshong brought up are addressed, this seems fine.
> 
> My thoughts on this are the following:
> 
> * I'm fairly impartial to the importing styles vs. global stylesheet discussion, but because this is a different pattern than we've used I think that it should be something that we all sign off on. Maybe throw a poll in the senior engineering channel?
> * Is there a way to set up module aliasing in SASS/Typescript to reduce the amount of pathing needed to import files? Think edio as a reference how we could import things by doing "@core/import/the-thing". It would be great if we could do that for the other stylesheets so that if a file is moved it doesn't regress a ton of components (a big bonus for the global stylesheet)
> 
> update: it looks like you can define module aliases in webpack for all things in general. [See here for reference.](https://webpack.js.org/configuration/resolve/)

@joshuapeters would you be cool with me breaking out the topic of "switching from typescript's classic module loading to aliasing" into another github issue?

**Comment by wintondeshong** [comment]
> While I'm fine with this direction if that's how the rest of the team would like to go, I'd probably prefer to keep it as-is, as it's what we typically do in projects, and I don't see a huge benefit in changing it. It feels a bit awkward to have to import the 'global' or 'base' classes in each file now - something we got for free before

Definitely more work as it comes to SCSS. It makes the SCSS feel more like the other import TS/JS files.

**Comment by wintondeshong** [comment]
> I'll go ahead and leave my two cents here. My optimistic side really likes this approach and the idea of keeping similar functionality close to each other (it lends well to vertical slice architecture concepts), however, my pessimistic side wonders it would be smart to at least find a way of "trying" this approach before committing to doing it across the board. Are we giving ourselves an unneeded maintenance burden if it turns out we don't like it? If it's not that big of deal to switch back, then I say, let's try it.

Thanks for weighing in @myty! Yeah, the idea is consumers have an easier path to selecting only the portions they need. where-as currently, while totally possible, consumers would need to buffet-style it and import only what they want from an outside-in approach.

I'm not sure if that point is entirely apparent to everyone. That our current approach _does_ support selectively including only what you want. It is just that the documentation's happy path states you import the 'global' 'andculture-javascript-react-components.scss' file. Consumers are totally cool to replace that file and pick and choose. If for some reason the group decides against this approach to put SCSS files beside components, perhaps the alternative is to update the documentation with this alternative to loading SCSS.

**Comment by jstevenson** [comment]
I like this approach, and it makes code cleanup a lot easier when dropping components (no more random CSS sitting in a file years after it was removed).

**Review by jstevenson — APPROVED** [review]

**Comment by wintondeshong** [comment]
@brandongregoryscott @dylanjustice @HeyKos @joshuapeters @jstevenson @KevinBusch @mrjones2014 @myty @Stefanie899 - It sounds like the majority of the group is in favor of moving forward with the move of component specific SCSS files (currently in `6-components/**/*`) be moved in parallel to their component file. 

@phess101 you should be good to roll forward with updating the rest of the components in the repository per this proposal.

At the end of the update, the `6-components` directory should just contain `_component-base.scss` (per the above)

**Comment by joshuapeters** [comment]
> > Assuming that all of the other points @wintondeshong brought up are addressed, this seems fine.
> > My thoughts on this are the following:
> > 
> > * I'm fairly impartial to the importing styles vs. global stylesheet discussion, but because this is a different pattern than we've used I think that it should be something that we all sign off on. Maybe throw a poll in the senior engineering channel?
> > * Is there a way to set up module aliasing in SASS/Typescript to reduce the amount of pathing needed to import files? Think edio as a reference how we could import things by doing "@core/import/the-thing". It would be great if we could do that for the other stylesheets so that if a file is moved it doesn't regress a ton of components (a big bonus for the global stylesheet)
> > 
> > update: it looks like you can define module aliases in webpack for all things in general. [See here for reference.](https://webpack.js.org/configuration/resolve/)
> 
> @joshuapeters would you be cool with me breaking out the topic of "switching from typescript's classic module loading to aliasing" into another github issue?

@wintondeshong I'm cool with that! Just to clarify, I'm also talking about NPM aliases to make the pathing much more concise in SASS files as well as Typescript files. The link I provided should allow us to accomplish both!

**Comment by phess101** [comment]
> @phess101 thanks for putting this proposal together! It does a good job of demonstrating how this works when applied to the rest of the repository's components.
> 
> There are a few discussions points below that'd I'd love to get your input on as well as other senior engineers. As for next steps, I'm going to share this PR with some other seniors to make sure we are all good with proceeding in this direction.
> 
> Once we get approval across the board, the next step will be for you to complete the PR by rolling it through the rest of the components. When that comes, please make sure to verify there aren't any warnings (right now there are quite a few) and that none of the components have visually or functionally regressed. Recommend doing a side-by-side with our current released version https://andculturecode.github.io/AndcultureCode.JavaScript.React.Components/?path=/story/*

@wintondeshong We are close to wrapping up this PR, but I was curious as to what warnings you are seeing that would be caused by these proposed changes? I'm going to go ahead and submit this PR for re-review in the meantime.

**Review by HeyKos — APPROVED** [review]
I am good with this approach. It is nice that all files related to a component are together. Thank you @TheDataDesigner and @phess101 for putting this PR together.

**Review by brandongregoryscott — APPROVED** [review]

File: package.json
Comment (phess101) [reply package.json:95]: Looks like VSCode changed the indent from 4 to 2 to be "helpful". Reverted back to 4 spaces for indent

Comment (brandongregoryscott) [reply package.json:95]: Weird - thanks for reverting it. I think one of our repos had a prettier config with a `tabWidth` value of 2 instead of 4 but was fixed up recently.

**Review by Stefanie899 — APPROVED** [review]

**Review by wintondeshong — APPROVED** [review]
@phess101 thank you for working in all of the feedback 👍

**Comment by brandongregoryscott** [comment]
@all-contributors add @phess101 for code

**Comment by brandongregoryscott** [comment]
@all-contributors add @alexgetty for code

**Comment by brandongregoryscott** [comment]
@all-contributors add @Stefanie899 for review

**Comment by brandongregoryscott** [comment]
@all-contributors add @HeyKos for review

**Comment by brandongregoryscott** [comment]
@all-contributors add @myty for review

**Comment by brandongregoryscott** [comment]
@all-contributors add @dylanjustice for review

**Comment by brandongregoryscott** [comment]
@all-contributors add @KevinBusch for review

**Comment by brandongregoryscott** [comment]
@all-contributors add @joshuapeters for review

**Pull request merged**
